### Student Activity: Indexing

This program reads daily stock data from a CSV, trims extraneous columns, sets the date as the index, and calculates and plots returns for a specified date range.

In [1]:
# Import libraries and dependencies
import pandas as pd
from pathlib import Path

### Read CSV in as DataFrame

In [2]:
# Set the path
file_path = Path('../Resources/loans.csv')

# Read in the CSV as a DataFrame
loans_csv = pd.read_csv(file_path)

### Show the First 10 Records

In [3]:
# Retrive rows with index 0 up to 10 (not including)
loans_csv.iloc[0:10]

loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  installment  \
0      10000        10000          10000.0  36 months    0.1033       324.23   
1       4000         4000           4000.0  36 months    0.2340       155.68   
2       5000         5000           5000.0  36 months    0.1797       180.69   
3       9600         9600           9600.0  36 months    0.1298       323.37   
4       2500         2500           2500.0  36 months    0.1356        84.92   
5      30000        30000          30000.0  60 months    0.1894       777.23   
6      23000        23000          23000.0  60 months    0.2089       620.81   
7      32075        32075          32075.0  60 months    0.1180       710.26   
8       8000         8000           8000.0  36 months    0.2340       311.35   
9      10000        10000          10000.0  60 months    0.1992       264.50   

  grade sub_grade           emp_title emp_length home_ownership  annual_inc  \
0     B        B1                 NaN   < 1 year       MORTGAGE    280000.0   
1     E        E1            Security    3 years           RENT     90000.0   
2     D        D1      Administrative    6 years       MORTGAGE     59280.0   
3     B        B5                 NaN        NaN       MORTGAGE     35704.0   
4     C        C1                Chef  10+ years           RENT     55000.0   
5     D        D2         Postmaster   10+ years       MORTGAGE     90000.0   
6     D        D4            Operator    5 years           RENT     68107.0   
7     B        B4  Nursing Supervisor  10+ years       MORTGAGE    150000.0   
8     E        E1             Manager  10+ years            OWN     43000.0   
9     D        D3    Material Handler  10+ years       MORTGAGE     80000.0   

  verification_status issue_d loan_status pymnt_plan             purpose  
0        Not Verified  Dec-18     Current          n  debt_consolidation  
1     Source Verified  Dec-18     Current          n  debt_consolidation  
2     Source Verified  Dec-18     Current          n  debt_consolidation  
3        Not Verified  Dec-18     Current          n    home_improvement  
4        Not Verified  Dec-18     Current          n  debt_consolidation  
5     Source Verified  Dec-18     Current          n  debt_consolidation  
6     Source Verified  Dec-18     Current          n  debt_consolidation  
7        Not Verified  Dec-18     Current          n         credit_card  
8     Source Verified  Dec-18     Current          n  debt_consolidation  
9        Not Verified  Dec-18     Current          n  debt_consolidation

### View Summary Statistics

In [8]:
# Describe summary statistics for csv datad
display(loans_csv.describe().T)
display(loans_csv.describe(include = 'object').T)

count          mean            std      min         25%  \
loan_amnt        128412.0  15971.321021   10150.384233  1000.00   8000.0000   
funded_amnt      128412.0  15971.321021   10150.384233  1000.00   8000.0000   
funded_amnt_inv  128412.0  15968.498166   10152.168970   725.00   8000.0000   
int_rate         128412.0      0.129302       0.050904     0.06      0.0846   
installment      128412.0    463.259353     285.718256    30.48    253.6300   
annual_inc       128412.0  82797.327861  108298.465792     0.00  47058.0000   

                       50%         75%           max  
loan_amnt        14000.000  21600.0000  4.000000e+04  
funded_amnt      14000.000  21600.0000  4.000000e+04  
funded_amnt_inv  14000.000  21600.0000  4.000000e+04  
int_rate             0.118      0.1614  3.099000e-01  
installment        382.905    622.6800  1.618240e+03  
annual_inc       68000.000  99000.0000  9.757200e+06

count unique                 top    freq
term                 128412      2           36 months   88179
grade                128412      7                   A   38011
sub_grade            128412     35                  A4    9770
emp_title            107465  43892             Teacher    2090
emp_length           116708     11           10+ years   38826
home_ownership       128412      4            MORTGAGE   63490
verification_status  128412      3        Not Verified   58350
issue_d              128412      3              Oct-18   46305
loan_status          128412      7             Current  119540
pymnt_plan           128412      1                   n  128412
purpose              128412     12  debt_consolidation   70603

### Filter Columns and Create Subset DataFrame

In [11]:
# Filter the DataFrame down to the following, keep all rows

#loan_amnt
#term
#int_rate
#emp_title
#annual_inc
#purpose

filtered_df = loans_csv.iloc[:, [0, 3, 4, 8, 11, 16]]
filtered_df.head()

loan_amnt       term  int_rate       emp_title  annual_inc  \
0      10000  36 months    0.1033             NaN    280000.0   
1       4000  36 months    0.2340        Security     90000.0   
2       5000  36 months    0.1797  Administrative     59280.0   
3       9600  36 months    0.1298             NaN     35704.0   
4       2500  36 months    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

### Filter Rows by 36 Month Terms

In [12]:
# Conditional indexing to filter DataFrame where 'term' is equal to '36 months'
term_df = filtered_df.loc[filtered_df['term'] == '36 months'].copy()
term_df.head()

loan_amnt       term  int_rate       emp_title  annual_inc  \
0      10000  36 months    0.1033             NaN    280000.0   
1       4000  36 months    0.2340        Security     90000.0   
2       5000  36 months    0.1797  Administrative     59280.0   
3       9600  36 months    0.1298             NaN     35704.0   
4       2500  36 months    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

### Modify Row Values

In [16]:
# Change row values within the 'term' column from '36 months' to '3 Years'
term_df.loc[term_df['term'] == '36 months', 'term'] = '3 Years'
term_df.head()

loan_amnt     term  int_rate       emp_title  annual_inc  \
0      10000  3 Years    0.1033             NaN    280000.0   
1       4000  3 Years    0.2340        Security     90000.0   
2       5000  3 Years    0.1797  Administrative     59280.0   
3       9600  3 Years    0.1298             NaN     35704.0   
4       2500  3 Years    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

In [15]:
term_df[term_df['term'] == '36 months']['term']
term_df.loc[term_df['term'] == '36 months', 'term']

0         36 months
1         36 months
2         36 months
3         36 months
4         36 months
            ...    
128407    36 months
128408    36 months
128409    36 months
128410    36 months
128411    36 months
Name: term, Length: 88179, dtype: object

In [17]:
# Change row values within the 'emp_title' column from NaN to 'Unknown'
term_df.loc[term_df['emp_title'].isnull(), 'emp_title'] = 'Unknown'
term_df.head()

loan_amnt     term  int_rate       emp_title  annual_inc  \
0      10000  3 Years    0.1033         Unknown    280000.0   
1       4000  3 Years    0.2340        Security     90000.0   
2       5000  3 Years    0.1797  Administrative     59280.0   
3       9600  3 Years    0.1298         Unknown     35704.0   
4       2500  3 Years    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

### View Summary Statistics for 3 Year Term Loans

In [18]:
# Describe summary statistics for three-year loans
term_df.describe(include='all')

loan_amnt     term      int_rate emp_title    annual_inc  \
count   88179.000000    88179  88179.000000     88179  8.817900e+04   
unique           NaN        1           NaN     31470           NaN   
top              NaN  3 Years           NaN   Unknown           NaN   
freq             NaN    88179           NaN     15669           NaN   
mean    13279.580456      NaN      0.120610       NaN  8.043616e+04   
std      9479.610905      NaN      0.048955       NaN  1.109251e+05   
min      1000.000000      NaN      0.060000       NaN  0.000000e+00   
25%      6000.000000      NaN      0.078400       NaN  4.500000e+04   
50%     10000.000000      NaN      0.110600       NaN  6.500000e+04   
75%     18000.000000      NaN      0.150200       NaN  9.500000e+04   
max     40000.000000      NaN      0.309900       NaN  9.000000e+06   

                   purpose  
count                88179  
unique                  12  
top     debt_consolidation  
freq                 45456  
mean                   NaN  
std                    NaN  
min                    NaN  
25%                    NaN  
50%                    NaN  
75%                    NaN  
max                    NaN

### Show Unique Values for Employee Titles with 3 Year Loans

In [20]:
# Calculate unique values and counts for employee titles of 3 year customer loans
term_df['emp_title'].value_counts(normalize = True)

Unknown                       0.177695
Teacher                       0.016251
Manager                       0.013688
Registered Nurse              0.007621
Driver                        0.007417
                                ...   
mgt dept. embroider           0.000011
Financial Services Officer    0.000011
Office Manager/Controller     0.000011
Motorcoach Operator           0.000011
Key Accounts Manager          0.000011
Name: emp_title, Length: 31470, dtype: float64

### Show Unique Values for Purpose of 3 Year Loans

In [11]:
# Calculate unique values and counts for loan purposes of 3 year customer loans
term_df['purpose'].value_counts()

debt_consolidation    45456
credit_card           24780
other                  5836
home_improvement       5446
major_purchase         1795
medical                1245
car                     838
small_business          828
vacation                749
house                   596
moving                  547
renewable_energy         63
Name: purpose, dtype: int64

### Show Summary Statistics for 3 Year Loans of Customers with Annual Income Greater Than $80,000

In [12]:
# Display summary statistics where annual income is greater than $80,000 to find count and mean
term_df.loc[term_df['annual_inc'] > 80000].describe()

loan_amnt      int_rate    annual_inc
count  29748.000000  29748.000000  2.974800e+04
mean   18240.116142      0.114116  1.389634e+05
std    10991.909117      0.048067  1.752361e+05
min     1000.000000      0.060000  8.000100e+04
25%    10000.000000      0.075600  9.500000e+04
50%    15700.000000      0.104700  1.120000e+05
75%    25000.000000      0.144700  1.500000e+05
max    40000.000000      0.309400  9.000000e+06

### Show Summary Statistics for 3 Year Loans of Customers with Annual Income Less Than $80,000

In [13]:
# Display summary statistics where annual income is less than $80,000 to find count and mean
term_df.loc[term_df['annual_inc'] < 80000].describe()

loan_amnt      int_rate    annual_inc
count  56015.000000  56015.000000  56015.000000
mean   10607.487727      0.124268  49372.805964
std     7318.791900      0.049137  16669.098821
min     1000.000000      0.060000      0.000000
25%     5000.000000      0.081900  37000.000000
50%     9600.000000      0.113100  50000.000000
75%    14000.000000      0.150200  62635.500000
max    40000.000000      0.309900  79999.000000

### What insights can we answer regarding 3 year loans?

1. What kind of customers (employee title) seem to ask for three-year loans most frequently?
2. What are three-year loans generally used for?
3. What is the difference in count of three-year loan customers with annual incomes greater than 80,000 compared to those with annual incomes less than 80,000?
4. What is the difference in interest rates of three-year loan customers with annual incomes greater than 80,000 compared to those with annual incomes less than 80,000? 

1. Interestingly enough Teachers and Managers seem to be the top two contenders for requesting three-year loans. In addition, there are a disproporionate amount of 'Unknown' or un-reported employee titles. Unless this field is optional, this could signal a potential data quality problem.

2. From the value counts of loan purposes, it looks like three-year loans are generally used to consolidate debt.

3. A majority of three-year loan customers have annual incomes of less than 80,000. Counts of customers with annual incomes less than 80,000 compared to greater than 80,000 stand at 56,015 customers vs 29,748 customers, respectively.

4. Interest rates between customers with annual incomes greater than 80,000 compared to those with less than 80,000 differ by approximately 1%; Interest rates were 12.43% and 12.41%, respectively.